### PR TITLE
[Framwork Add]Add examples

### PR DESCRIPTION
**Pull request by Jim8y**
This pr adds contract examples.

### CODE REVIEW

**Comment by Jim8y** [comment]
@shargon no need to worry about this pr, let @chenzhitong handle it.

File: neo-devpack-dotnet.sln [hunk 1]
```diff
@@ -37,10 +37,37 @@ EndProject
...
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Neo.SmartContract.Template.UnitTests", "tests\Neo.SmartContract.Template.UnitTests\Neo.SmartContract.Template.UnitTests.csproj", "{17F45E0B-AB1C-4796-8C99-E5212A5592F8}"
+EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Neo.Extensions", "neo\src\Neo.Extensions\Neo.Extensions.csproj", "{E5EFB018-810D-4297-8921-940FA0B1ED97}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Neo.IO", "neo\src\Neo.IO\Neo.IO.csproj", "{C2B7927F-AAA5-432A-8E76-B5080BD7EFB9}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Examples", "Examples", "{C10F9957-077F-42C8-B350-8607D4899B7E}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Example.SmartContract.ContractCall", "examples\Example.SmartContract.ContractCall\Example.SmartContract.ContractCall.csproj", "{AE9BF707-FA85-41CB-BE04-3B0AD024329B}"
```
Comment (shargon) [inline neo-devpack-dotnet.sln:47]: it's possible to include only the exaples.sln?

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_StaticVar.cs [hunk 2]
```diff
@@ -24,20 +23,20 @@ public class Contract_StaticVar : SmartContract.Framework.SmartContract
...
         /// </summary>
-        [InitialValue("NXV7ZhHiyM1aHXwpVsRZC6BwNFP2jghXAq", ContractParameterType.Hash160)]
+        [Hash160("NXV7ZhHiyM1aHXwpVsRZC6BwNFP2jghXAq")]
         private static readonly UInt160 uInt160 = default;
 
         /// <summary>
         /// A static string field initialized with the InitialValue attribute. This demonstrates initializing contract fields that cannot be directly assigned with their value at compile time.
         /// </summary>
-        [InitialValue("hello world", ContractParameterType.String)]
+        [String("hello world")]
```
[the remark this answers is not in the document]
Comment (Jim8y) [reply tests/Neo.Compiler.CSharp.TestContracts/Contract_StaticVar.cs:39]: Maybe next pr, and you can do it. This pr will only focus on update.

File: neo-devpack-dotnet.sln (hunk 1 above)
Comment (Jim8y) [reply neo-devpack-dotnet.sln:47]: Currently no, i am not familier with this, but removing them will cause the solution not able to find examples.

**Comment by Jim8y** [comment]
Lets go then. @shargon please review again.

File: neo-devpack-dotnet.sln (hunk 1 above)
Comment (shargon) [reply neo-devpack-dotnet.sln:47]: We should archive https://github.com/neo-project/examples ?

File: examples/Example.SmartContract.Event/Example.SmartContract.Event.csproj [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net7.0</TargetFramework>
```
Comment (shargon) [inline examples/Example.SmartContract.Event/Example.SmartContract.Event.csproj:4]: Remove framework for all the projects? it's duplicated with directory.build.props

Comment (Jim8y) [reply examples/Example.SmartContract.Event/Example.SmartContract.Event.csproj]: Addressed

File: neo-devpack-dotnet.sln (hunk 1 above)
Comment (Jim8y) [reply neo-devpack-dotnet.sln:47]: I am not going to maintain that one anyway. I want to have the latest compiler features.

**Comment by shargon** [comment]
Could you add to workflow the compilation of all of these tests, just to ensure that they compile good, now it should be easier with multiple sources, but maybe we need to be able to compile a .sln...

**Comment by Jim8y** [comment]
dude, if it can be fixed in another pr, just allow me to do it in another pr. now conflicts after conflicts again.

**Comment by Jim8y** [comment]
> Could you add to workflow the compilation of all of these tests? just to ensure that they compile good, now it should be easier with multiple sources, but maybe we need to be able to compile a .sln... (or multiple csproj)

This is not multiple source, but multiple projects, and they do compile as when you compile the solution, they compile:

Build with surface heuristics started at 17:19:07
Use build tool: C:\Program Files\dotnet\sdk\7.0.406\MSBuild.dll
CONSOLE: MSBuild version 17.7.4+3ebbd7c49 for .NET
CONSOLE: Build started 2/26/2024 5:19:07 PM.
CONSOLE: Project "C:\Users\liaoj\AppData\Local\Temp\Vazuwur.proj" on node 1 (default targets).
CONSOLE: ControllerTarget:
CONSOLE:   Run controller from C:\Program Files\JetBrains\Rider\r2r\2023.3.3R\772133E8377C021858F9D863C63ADAC\JetBrains.Platform.MsBuildTask.v17.dll
9>------- Started building project: Neo.SmartContract.Template
10>------- Started building project: Neo.Compiler.CSharp.TestContracts
11>------- Started building project: Neo.SmartContract.Framework.TestContracts
12>------- Started building project: Example.SmartContract.ContractCall
13>------- Started building project: Example.SmartContract.Event
14>------- Started building project: Example.SmartContract.Exception
15>------- Started building project: Example.SmartContract.HelloWorld
16>------- Started building project: Example.SmartContract.Inscription
17>------- Started building project: Example.SmartContract.Modifier
18>------- Started building project: Example.SmartContract.NEP17
19>------- Started building project: Example.SmartContract.NFT
20>------- Started building project: Example.SmartContract.Oracle
21>------- Started building project: Example.SmartContract.Storage
22>------- Started building project: Example.SmartContract.Transfer
23>------- Started building project: Example.SmartContract.ZKP

File: tests/Neo.SmartContract.Framework.TestContracts/Contract_SupportedStandard11Enum.cs [hunk 4]
```diff
@@ -4,8 +4,9 @@
 
 namespace Neo.SmartContract.Framework.UnitTests.TestClasses
 {
+
```
Comment (shargon) [inline tests/Neo.SmartContract.Framework.TestContracts/Contract_SupportedStandard11Enum.cs:7]:
```suggestion
```

File: src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs [hunk 5]
```diff
@@ -37,10 +37,16 @@ public static UInt160 GetOwner()
         [DisplayName("SetOwner")]
         public static event OnSetOwnerDelegate OnSetOwner;
 
+<<<<<<< HEAD:src/Neo.SmartContract.Template/templates/neocontractowner/Contract1.cs
```
Comment (shargon) [inline src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs:40]: This compile? :S

Comment (Jim8y) [reply src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs:40]: sh**

File: examples/Directory.Build.props [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        <IncludeSymbols>true</IncludeSymbols>
+        <SymbolPackageFormat>snupkg</SymbolPackageFormat>
+        <Company>The Neo N3 Contract Examples Project</Company>
+        <GenerateDocumentationFile>true</GenerateDocumentationFile>
+        <Nullable>enable</Nullable>
+        <AllowUnsafeBlocks>false</AllowUnsafeBlocks>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <ProjectReference Include="..\..\src\Neo.SmartContract.Framework\Neo.SmartContract.Framework.csproj" />
```
Comment (shargon) [inline examples/Directory.Build.props:24]: This works with the current compiler?

Comment (Jim8y) [reply examples/Directory.Build.props:24]: Not compiled, just syntax

**Comment by Jim8y** [comment]
BTW, @shargon you might need to fix the template compile issue, that is what caused the build issue in the workflow of this pr.

File: examples/Directory.Build.props (hunk 6 above)
Comment (shargon) [reply examples/Directory.Build.props:24]: I mean, in `public Compilation GetCompilation(string csproj)` we don't read the Directory.Build.props, maybe we should do it, (not related to this pr)

File: src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs (hunk 5 above)
Comment (shargon) [reply src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs]: > Yes, it compiles without error.

Maybe it's related to https://github.com/neo-project/neo-devpack-dotnet/pull/908#issuecomment-1963088333

File: examples/Directory.Build.props (hunk 6 above)
Comment (Jim8y) [reply examples/Directory.Build.props:24]: emmmm, to be honest, i never thought we will run into this situation

File: src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs (hunk 5 above)
Comment (Jim8y) [reply src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs]: > > Yes, it compiles without error.
> 
> Maybe it's related to [#908 (comment)](https://github.com/neo-project/neo-devpack-dotnet/pull/908#issuecomment-1963088333)

I dont get it, you have unit test in there, if its not built, how could the test run, but if you build it, why no error?

Comment (Jim8y) [reply src/Neo.SmartContract.Template/templates/neocontractowner/Ownable.cs]: There is something wrong with the templates, syntax analyzer not working on it at all

**Comment by Jim8y** [comment]
Man, please do it, i am scared of conflicts here~~~

**Comment by Jim8y** [comment]
BTW, @shargon please check the examples, and help to polish it either in this pr or in another pr. better in another pr.....

**Comment by shargon** [comment]
> BTW, @shargon please check the examples, and help to polish it either in this pr or in another pr. better in another pr.....

Check now, template pr was merged

**Comment by shargon** [comment]
> Man, please do it, i am scared of conflicts here~~~

two files only :)

**Comment by Jim8y** [comment]
@shargon  https://github.com/neo-project/neo-devpack-dotnet/actions/runs/8062085901/job/22021126211?pr=931 here it is, please check this build error. You need to fix this.

File: src/Neo.SmartContract.Framework/Interfaces/INEP11Payment.cs [hunk 7]
```diff
@@ -1,3 +1,4 @@
+#nullable enable
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Interfaces/INEP11Payment.cs:1]: We can make nullable the framework in a different pr

File: tests/Neo.SmartContract.Template.UnitTests/templates/neocontractoracle/OracleRequestTests.cs [hunk 8]
```diff
@@ -6,8 +6,9 @@
 using Neo.VM;
 using System.Text;
 
-namespace Neo.SmartContract.Template.UnitTests.templates.neocontractowner
+namespace Neo.SmartContract.Template.UnitTests.templates.neocontractoracle
 {
+
```
Comment (shargon) [inline tests/Neo.SmartContract.Template.UnitTests/templates/neocontractoracle/OracleRequestTests.cs:11]:
```suggestion
```

File: examples/Example.SmartContract.ZKP/ZKP.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+using System;
+using System.ComponentModel;
+
+namespace ZKP
+{
+    [DisplayName("SampleZKP")]
+    [ContractAuthor("code-dev")]
+    [ContractVersion("0.0.1")]
+    [ContractDescription("A sample contract to demonstrate how to use Example.SmartContract.ZKPil")]
+    [ContractSourceCode("https://github.com/neo-project/neo-devpack-dotnet/examples/ZKP")]
```
Comment (shargon) [inline examples/Example.SmartContract.ZKP/ZKP.cs:24]: this is not the write place

File: examples/Example.SmartContract.Transfer/TransferContract.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+namespace Transfer;
+
+/// <summary>
+/// This is a sample contract that can be used as a template for creating new contracts.
+/// </summary>
+[DisplayName("Contract1")]
+[ContractAuthor("<Your Name Or Company Here>", "<Your Public Email Here>")]
+[ContractDescription("<Description Here>")]
+[ContractVersion("1.0.0.0")]
+[ContractSourceCode("https://github.com/cschuchardt88/neo-templates")]
```
Comment (shargon) [inline examples/Example.SmartContract.Transfer/TransferContract.cs:30]: wrong url, maybe all wrong

File: src/Neo.SmartContract.Framework/Interfaces/INEP11Payment.cs (hunk 7 above)
Comment (Jim8y) [inline src/Neo.SmartContract.Framework/Interfaces/INEP11Payment.cs:1]:
```suggestion
```

File: examples/Example.SmartContract.Transfer/TransferContract.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+using System.Numerics;
+using Neo;
+using Neo.SmartContract.Framework.Services;
+
+namespace Transfer;
+
+/// <summary>
+/// This is a sample contract that can be used as a template for creating new contracts.
+/// </summary>
+[DisplayName("Contract1")]
```
Comment (shargon) [inline examples/Example.SmartContract.Transfer/TransferContract.cs:26]: Also we should use a different name, or remove the attribute and take the class name

File: examples/Example.SmartContract.ZKP/ZKP.cs (hunk 9 above)
Comment (Jim8y) [reply examples/Example.SmartContract.ZKP/ZKP.cs:24]: what?

Comment (shargon) [reply examples/Example.SmartContract.ZKP/ZKP.cs:24]: /examples/Example.SmartContract.ZKP/ZKP.cs

File: examples/Example.SmartContract.Transfer/TransferContract.cs (hunk 11 above)
Comment (Jim8y) [reply examples/Example.SmartContract.Transfer/TransferContract.cs]: fixed

File: examples/Example.SmartContract.ZKP/ZKP.cs (hunk 9 above)
Comment (Jim8y) [reply examples/Example.SmartContract.ZKP/ZKP.cs]: no more detailed, but all goes to examples.

File: examples/Example.SmartContract.ContractCall/ContractCall.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+using Neo.SmartContract.Framework.Attributes;
+using Neo.SmartContract.Framework.Services;
+using System.ComponentModel;
+using System.Numerics;
+
+namespace ContractCall;
+
+[DisplayName("SampleContractCall")]
+[ContractAuthor("core-dev")]
+[ContractEmail("core@neo.org")]
```
Comment (shargon) [inline examples/Example.SmartContract.ContractCall/ContractCall.cs:22]: is not `dev@neo.org`?

Comment (Jim8y) [reply examples/Example.SmartContract.ContractCall/ContractCall.cs:22]: I think its example. should be fine.

File: examples/Example.SmartContract.Oracle/Oracle.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace Oracle
+{
+    [DisplayName("SampleOracle")]
+    [ContractAuthor("code-dev", "core@neo.org")]
+    [ContractDescription("A sample contract to demonstrate how to use Example.SmartContract.Oracle Service")]
+    [ContractVersion("0.0.1")]
+    [ContractSourceCode("https://github.com/neo-project/neo-devpack-dotnet/tree/master/examples/")]
+    [ContractPermission(Permission.WildCard, Method.WildCard)]
+    public class SampleOracle : SmartContract
```
Comment (shargon) [inline examples/Example.SmartContract.Oracle/Oracle.cs:27]: Does we need to replicate the template?

**Comment by shargon** [comment]
I think that we should have unit tests for the examples that we provide, this will ensure the good practice to devs, in my opinion it's better to have less samples, but well tested and documented

**Review by shargon — APPROVED** [review]
@Jim8y We can improve it lated, I think that it's better to move them to templates, and be tested

**Comment by Jim8y** [comment]
I can do it later, can't wait to merge it, your turn to fix conflicts!!!!!!!!!
